src/operation.jl

```diff
@@ -56,9 +60,18 @@ for FUN in (:applyeager, :applylazy, :applypermute,
             param = randparam(op, imgs)
             map(img -> ($FUN)(op, img, param), imgs)
```

Comment:
I'm trying to wrap my head around the implementation. Before this commit, `imgs` contained images of the same datatype, after, it contains AbstractArrays and Masks. So inside the call to `map()` dynamic dispatch is triggered. Is this a problem? I think this could be avoided somehow by changing codegen.jl to specialize / optimize the given pipeline twice: once for the image, once for the masks. Or is this already happening with this implementation? All the `augment_impl` code looks a bit like black magic to me.

Reply:
> So inside the call to map() dynamic dispatch is triggered.

No, it won't. That's why I introduced `Val{false}` and `Val{true}` so that the `true`/`false` information can be passed to Julia at compiled time.

Reply:
Also we can benchmark this:

```julia
using Augmentor
img = Augmentor.testpattern()

@btime augment(($img, $img), Scale(0.5));
# this PR: 1.262 ms (60 allocations: 243.34 KiB)
# master: 1.318 ms (60 allocations: 243.34 KiB)

@btime augment(($img, $(Augmentor.Mask(img))), Scale(0.5));
# 1.262 ms (60 allocations: 243.34 KiB)

Reply:
So the map-over-tuple seems to be unrolled during compilation, nice :)

Reply:
If there are runtime dispatch, usually there are suspicious memory allocations, e.g.,

```julia
julia> rand01_stable() = rand() > 0.5 ? 0 : 1
rand01_stable (generic function with 1 method)

julia> rand01_unstable() = rand() > 0.5 ? 0.0 : 1
rand01_unstable (generic function with 1 method)

julia> g(f) = [f() for _ in 1:100]
g (generic function with 2 methods)

julia> @btime g(rand01_stable);
  182.425 ns (1 allocation: 896 bytes)

julia> @btime g(rand01_unstable);
  1.031 μs (46 allocations: 2.44 KiB)
```

Another new tool that I haven't yet understand well is JETTest

```julia
julia> @report_dispatch augment((img, img), Scale(0.5))
No errors !
(Tuple{Matrix{ColorTypes.RGBA{FixedPointNumbers.N0f8}}, Matrix{ColorTypes.RGBA{FixedPointNumbers.N0f8}}}, 0)
julia> @report_dispatch augment((img, Augmentor.Mask(img)), Scale(0.5))
No errors !
(Tuple{Matrix{ColorTypes.RGBA{FixedPointNumbers.N0f8}}, Augmentor.Mask{Matrix{ColorTypes.RGBA{FixedPointNumbers.N0f8}}}}, 0)
```

Reply:
> So the map-over-tuple seems to be unrolled during compilation, nice :)

Yes. It's actually `Tuple{typeof(img1), typeof(img2), ...}` so Julia fully get the type information to do dispatch.

Reply:
Was it actually required that `imgs` contained images of the same datatype?

> So inside the call to map() dynamic dispatch is triggered (?).

Yes. For ordinary images, the methods that already were there (`applyeager`, `applylazy`, etc.) are called -- so nothing changes in that case. For semantic wrappers [these methods](https://github.com/Evizero/Augmentor.jl/pull/87/files#diff-0ca4f4ab4d56ca81cfe83d68e80324a477d924133b2978b58c0cdf6f8b89b8c8R69) are called, which then determine if the given operation should be applied and use [dispatching](https://github.com/Evizero/Augmentor.jl/pull/87/files#diff-0ca4f4ab4d56ca81cfe83d68e80324a477d924133b2978b58c0cdf6f8b89b8c8R70-R71) once more to apply/skip the operation. I am not sure if any of that is a problem.

> changing codegen.jl to specialize / optimize the given pipeline twice: once for the image, once for the masks

It is hard for me to understand the code in `codegen.jl` too. However, I believe that the pipelines are optimized w.r.t. the included operations, not the input. Also, it would perhaps require bigger changes to optimize the pipeline separately for images and for masks because we need to apply the same operations on both images and masks (if applicable) so that the respective pixels still correspond after the augmentation. So the rough approach is
```
for operation in pipeline
    produce operation's parameters
    apply operation on each input
```
not
```
for image in input
    augment image using all operations in pipeline
```

Reply:
>  I believe that the pipelines are optimized w.r.t. the included operations, not the input.

Correct. The idea is to collapse multiple operations into one so that computation goes faster. Say,

- (affine optimization): applying two coordinate transformations can be collapsed into composing two transformation matrices (which is at most 3*3 matrix multiplication)
- (lazy): applying two lazy operations can be done via broadcasting so that we can reduce intermediate memory allocation.

This `Mask` decorates the input data, which is not what codegen currently cares about.

> Also, it would perhaps require bigger changes to optimize the pipeline separately for images and for masks

Because NoOp has the best property that we can take advantage of. One possible way/hack is to add one more pass to change unactivated operations to NoOp before the current `code_gen`. With this extra pass, [`seek_connected` can get better results to optimize](https://github.com/Evizero/Augmentor.jl/blob/3c62635457395def90c63690847aa8a26d104b49/src/codegen.jl#L50-L52).